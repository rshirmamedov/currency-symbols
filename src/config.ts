/** Used in setTimeout to mimic async behavior */
export const TIMEOUT = 3000;

export const LOGIN_INCORRECT_EMAIL = 'incorrect@email.com';
export const LOGIN_INCORRECT_PASSWORD = 'incorrect-password';

export const API_BASE_URL = 'https://653fb0ea9e8bd3be29e10cd4.mockapi.io/api/v1';