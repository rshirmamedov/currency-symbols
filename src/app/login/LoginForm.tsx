"use client";
import { Button } from "@/components/Button";
import { Field } from "@/components/Field";
import Link from "next/link";
import { useForm } from "react-hook-form";
import { validateEmail, validatePassword } from "./validations";
import { sleep } from "@/utils";
import { TIMEOUT } from "@/config";
import type { LoginFormFields } from './types';

type LoginFormProps = {
    onSubmit?: () => void;
};

export const LoginForm = (props: LoginFormProps) => {
    const {
        register,
        handleSubmit,
        formState: { errors, isSubmitting },
    } = useForm<LoginFormFields>();

    const onSubmit = async () => {
        await sleep(TIMEOUT);
        props.onSubmit?.();
    };

    return (
        <div data-testid="login-form" className="max-w-md w-full p-8 border border-gray-300 rounded-md bg-white">
            <h1 className="text-center text-2xl">Login</h1>
            <div>
                <form onSubmit={handleSubmit(onSubmit)}>
                    <Field
                        inputId="email"
                        type="text"
                        label="Email"
                        placeholder="Enter email"
                        disabled={isSubmitting}
                        error={errors.email?.message}
                        {...register("email", { validate: validateEmail })}
                    />
                    <Field
                        inputId="password"
                        type="password"
                        label="Password"
                        placeholder="Enter password"
                        disabled={isSubmitting}
                        error={errors.password?.message}
                        {...register("password", { validate: validatePassword })}
                    />
                    <Button
                        data-testid="login-button"
                        title={isSubmitting ? "Logging in..." : "Login"}
                        disabled={isSubmitting}
                    />
                </form>
            </div>
            <div className="text-center p-4 text-sm ">
                <span className="text-blue-600 hover:underline">
                    <Link href="/registration">Register</Link>
                </span>
            </div>
        </div>
    );
};
