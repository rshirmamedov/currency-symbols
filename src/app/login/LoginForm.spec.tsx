import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { LoginForm } from "./LoginForm";
import { LOGIN_INCORRECT_EMAIL, LOGIN_INCORRECT_PASSWORD } from "@/config";

describe("LoginForm", () => {
    it("renders required email and password errors", async () => {
        render(<LoginForm />);

        await userEvent.click(await screen.findByTestId("login-button"));

        expect(await screen.findByTestId('email-field')).toMatchSnapshot();
        expect(await screen.findByTestId('password-field')).toMatchSnapshot();
    });

    it("renders invalid email error", async () => {
        const user = userEvent.setup();

        render(<LoginForm />);

        const emailInput = screen.getByLabelText(/email/i);
        const passwordInput = screen.getByLabelText(/password/i);

        await user.type(emailInput, 'john');
        await user.type(passwordInput, 'password');

        await userEvent.click(await screen.findByTestId("login-button"));

        expect(await screen.findByTestId('email-field')).toMatchSnapshot();
    });

    it("renders incorrect email error", async () => {
        const user = userEvent.setup();

        render(<LoginForm />);

        const emailInput = screen.getByLabelText(/email/i);
        const passwordInput = screen.getByLabelText(/password/i);

        await user.type(emailInput, LOGIN_INCORRECT_EMAIL);
        await user.type(passwordInput, 'password');

        await userEvent.click(await screen.findByTestId("login-button"));

        expect(await screen.findByTestId('email-field')).toMatchSnapshot();
    });

    it("renders incorrect password error", async () => {
        const user = userEvent.setup();

        render(<LoginForm />);

        const emailInput = screen.getByLabelText(/email/i);
        const passwordInput = screen.getByLabelText(/password/i);

        await user.type(emailInput, 'john@doe.com');
        await user.type(passwordInput, LOGIN_INCORRECT_PASSWORD);

        await userEvent.click(await screen.findByTestId("login-button"));

        expect(await screen.findByTestId('password-field')).toMatchSnapshot();
    });

    it("disables login button while submitting", async () => {
        const user = userEvent.setup();

        render(<LoginForm />);

        const emailInput = screen.getByLabelText(/email/i);
        const passwordInput = screen.getByLabelText(/password/i);

        await user.type(emailInput, 'john@doe.com');
        await user.type(passwordInput, "test123");

        await userEvent.click(await screen.findByTestId("login-button"));

        expect(await screen.findByTestId('login-button')).toMatchSnapshot();
    });

});