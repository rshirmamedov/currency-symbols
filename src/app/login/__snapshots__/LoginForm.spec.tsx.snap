// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm disables login button while submitting 1`] = `
<button
  class="w-full py-2.5 px-4 text-sm rounded text-white bg-blue-600 hover:bg-blue-700 disabled:bg-blue-400 focus:outline-none mt-6"
  data-testid="login-button"
  disabled=""
>
  Logging in...
</button>
`;

exports[`LoginForm renders incorrect email error 1`] = `
<div
  class="py-2"
  data-testid="email-field"
>
  <label
    class="block text-sm font-medium leading-6 text-gray-900"
    for="email"
  >
    Email
  </label>
  <input
    class="w-full text-sm p-3 rounded outline-none border-2 focus:border-blue-500"
    id="email"
    name="email"
    placeholder="Enter email"
    type="text"
  />
  <div
    class="pt-2"
  >
    <div
      class="text-sm text-red-600"
    >
      Email is incorrect
    </div>
  </div>
</div>
`;

exports[`LoginForm renders incorrect password error 1`] = `
<div
  class="py-2"
  data-testid="password-field"
>
  <label
    class="block text-sm font-medium leading-6 text-gray-900"
    for="password"
  >
    Password
  </label>
  <input
    class="w-full text-sm p-3 rounded outline-none border-2 focus:border-blue-500"
    id="password"
    name="password"
    placeholder="Enter password"
    type="password"
  />
  <div
    class="pt-2"
  >
    <div
      class="text-sm text-red-600"
    >
      Password is incorrect
    </div>
  </div>
</div>
`;

exports[`LoginForm renders invalid email error 1`] = `
<div
  class="py-2"
  data-testid="email-field"
>
  <label
    class="block text-sm font-medium leading-6 text-gray-900"
    for="email"
  >
    Email
  </label>
  <input
    class="w-full text-sm p-3 rounded outline-none border-2 focus:border-blue-500"
    id="email"
    name="email"
    placeholder="Enter email"
    type="text"
  />
  <div
    class="pt-2"
  >
    <div
      class="text-sm text-red-600"
    >
      Email is invalid
    </div>
  </div>
</div>
`;

exports[`LoginForm renders required email and password errors 1`] = `
<div
  class="py-2"
  data-testid="email-field"
>
  <label
    class="block text-sm font-medium leading-6 text-gray-900"
    for="email"
  >
    Email
  </label>
  <input
    class="w-full text-sm p-3 rounded outline-none border-2 focus:border-blue-500"
    id="email"
    name="email"
    placeholder="Enter email"
    type="text"
  />
  <div
    class="pt-2"
  >
    <div
      class="text-sm text-red-600"
    >
      Email is required
    </div>
  </div>
</div>
`;

exports[`LoginForm renders required email and password errors 2`] = `
<div
  class="py-2"
  data-testid="password-field"
>
  <label
    class="block text-sm font-medium leading-6 text-gray-900"
    for="password"
  >
    Password
  </label>
  <input
    class="w-full text-sm p-3 rounded outline-none border-2 focus:border-blue-500"
    id="password"
    name="password"
    placeholder="Enter password"
    type="password"
  />
  <div
    class="pt-2"
  >
    <div
      class="text-sm text-red-600"
    >
      Password is required
    </div>
  </div>
</div>
`;
